### ISSUE TITLE
New Generic syntax is not covariant

### ISSUE THREAD

**Issue by ilius** [astral-sh/ty#488]
### Summary

Before Python 3.12, we would write this

```py
from typing import Generic, TypeVar

T_co = TypeVar("T_co", covariant=True)

class A(Generic[T_co]):
    def __init__(self, value: T_co):
        pass


a: A[int | str] = A(1)
```

But starting from 3.12, this is equivalent:
```
class A[T_co]:
    def __init__(self, value: T_co):
        pass


a: A[int | str] = A(1)
```

The type should be covariant according to https://peps.python.org/pep-0695/

But `ty` will say:
```
error[invalid-assignment]: Object of type `A[int]` is not assignable to `A[int | str]`
 --> ty-generic.py:6:1
  |
6 | a: A[int | str] = A(1)
  | ^
  |
info: rule `invalid-assignment` is enabled by default

Found 1 diagnostic
```

**Comment by dcreager** [comment]
Variance inference is a known missing feature, but we did not yet have an issue to track that. So thank you for reporting this!

**Comment by ericmarkmartin** [comment]
I plan to work on this!

**Comment by InSyncWithFoo** [comment]
One minor thing that we might need to look out for is "private" attributes:

```python
class C[T]:
    def __init__(self, val: T) -> None:
        self._val = val  # _val: T
```

[Mypy, Pyright and Pyrefly all say that `T` is covariant](https://stackoverflow.com/a/79695291), since they consider `_val` to be non-public. Change that to `val`, and `T` becomes invariant.

**Comment by sharkdp** [comment]
I'm closing https://github.com/astral-sh/ty/issues/158, so I'm adding a note here that attributes declared `Final` might affect variance inference (covariant instead of invariant). The information whether a attribute is `Final` or not should already be available in the corresponding `qualifiers: TypeQualifiers` field when accessing a member on a class.
